### PR TITLE
Filter state, events_before and events_after in /context requests

### PR DESCRIPTION
**Pull request by babolivier**
While the current version of the spec doesn't say much about how this endpoint uses filters (see matrix-org/matrix-doc#2338), the current implementation is that some fields of an `EventFilter` apply (the ones that are used when running the SQL query) and others don't (the ones that are used by the filter itself) because we don't call `event_filter.filter(...)`. This seems counter-intuitive and probably not what we want (as per the matrix-doc issue and https://github.com/matrix-org/matrix-doc/pull/2344#discussion_r342449974) so this PR fixes it.

It also adds a bunch of tests to test filtering on `/context` and reorganises the labelling tests in `room.py` to minimise code duplication.

It also adds tests for using `/search` with a filter, since it's another endpoint the spec says should accept being given a filter, and update the copyright comments for the files edited by #6301.

Commits should be reviewable independently.

### CODE REVIEW

**Comment by richvdh** [comment]
> Commits should be reviewable independently.

maybe, but there's 13 of the things!

File: synapse/handlers/room.py [hunk 1]
```diff
@@ -902,7 +907,13 @@ def filter_evts(events):
         state = yield self.state_store.get_state_for_events(
             [last_event_id], state_filter=state_filter
         )
-        results["state"] = list(state[last_event_id].values())
+
+        # Apply the filter on state events.
```
Comment (richvdh) [inline synapse/handlers/room.py]: I don't think this is adding much value.
```suggestion
```

File: synapse/handlers/room.py [hunk 2]
```diff
@@ -902,7 +907,13 @@ def filter_evts(events):
...
+        state_events = list(state[last_event_id].values())
+        if event_filter:
+            state_events = event_filter.filter(state_events)
+
+        results["state"] = list(state_events)
```
Comment (richvdh) [inline synapse/handlers/room.py]: I don't think this second `list` is necessary?
```suggestion
        results["state"] = state_events
```

File: synapse/api/filtering.py [hunk 3]
```diff
@@ -1,5 +1,8 @@
 # -*- coding: utf-8 -*-
 # Copyright 2015, 2016 OpenMarket Ltd
+# Copyright 2017 Vector Creations Ltd
```
[the remark this answers is not in the document]
Comment (babolivier) [reply synapse/api/filtering.py]: fwiw, that PR was open soon after #6301 got merged, in which I forgot to update copyright headers, so I decided to update the headers of the files changed by #6301 in this PR.

**Pull request merged**